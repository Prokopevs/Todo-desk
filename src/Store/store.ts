import editModeSlice from './reducers/editMode/slice';
import { configureStore } from '@reduxjs/toolkit'
import createSagaMiddleware from 'redux-saga'
import authorizationSlice from './reducers/authorization/slice'
import dndSlice from './reducers/dnd/slice'
import prioritySlice from './reducers/priority/slice';
import errorMessageSlice from './reducers/errorMessage/slice';
import rootSaga from './Sagas';

const saga = createSagaMiddleware()

export const store = configureStore({
    reducer: {
        authorizationSlice,
        dndSlice,
        prioritySlice,
        editModeSlice,
        errorMessageSlice,
    },
    middleware: [saga]
})
saga.run(rootSaga)

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch