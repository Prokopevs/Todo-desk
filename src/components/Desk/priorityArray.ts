const priorityArray = [
    { color: "purple", description: "Very High" },
    { color: "red", description: "High" },
    { color: "green", description: "Normal" },
    { color: "blue", description: "Low" },
    { color: "lightblue", description: "Very Low" },
]

export default priorityArray;