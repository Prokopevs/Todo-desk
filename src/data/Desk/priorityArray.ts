const priorityArray = [
    { color: "purple", description: "Very High", index: 0 },
    { color: "red", description: "High", index: 1 },
    { color: "green", description: "Normal", index: 2 },
    { color: "blue", description: "Low", index: 3 },
    { color: "lightblue", description: "Very Low", index: 4 },
]

export default priorityArray