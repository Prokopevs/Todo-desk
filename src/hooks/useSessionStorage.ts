import { useState, useEffect } from "react";
 
function getStorageValue(key, defaultValue) {
    // getting stored value
    if (typeof window !== "undefined") {
        const saved = sessionStorage.getItem(key);
        const initial = saved !== null ? JSON.parse(saved) : defaultValue;
        return initial;
    }
}
 
export const useSessionStorage = (key, defaultValue) => {
    const [value, setValue] = useState(() => {
    return getStorageValue(key, defaultValue);
    });
 
    useEffect(() => {
        // storing input name
        sessionStorage.setItem(key, JSON.stringify(value));
    }, [key, value]);
 
    return [value, setValue];
};