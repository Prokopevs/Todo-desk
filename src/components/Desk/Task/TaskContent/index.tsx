import React from "react"
import { ITasksContent } from "../../../../models/Task/ITasksContent"
import TextareaAutosize from "react-textarea-autosize"
import { useForm, SubmitHandler } from "react-hook-form"
import { useAppDispatch, useAppSelector } from "../../../../hooks/redux"
import {
    changeTaskContent,
    changeTaskContentQuery,
} from "../../../../Store/reducers/dndSlice"
import TasksContentEditMode from "./TaskContentEditMode"
import {
    deleteTaskInEditArray,
    deleteTaskInSuccessArray,
} from "../../../../Store/reducers/editModeSlice"
import { motion, AnimatePresence } from "framer-motion"

interface Inputs {
    id: string
    content: string
    priority: number
    status_id: string
}

const TasksContent: React.FC<ITasksContent> = ({
    task,
    editMode,
    priorityArray,
    column,
    setEditMod,
}) => {
    const { priority } = useAppSelector((state) => state.dndSlice.data.tasks[task.id])
    const { isAuth } = useAppSelector((state) => state.authorizationSlice)
    const { prevTaskObj, successTasksAfterSagaRequest } = useAppSelector(
        (state) => state.editModeSlice
    )
    const dispatch = useAppDispatch()

    const moveCaretAtEnd = (e) => {
        let temp_value = e.target.value
        e.target.value = ""
        e.target.value = temp_value
    }

    const {
        register,
        handleSubmit,
        formState: { errors, isDirty, isValid },
    } = useForm<Inputs>({ mode: "onChange" })

    const onSubmit: SubmitHandler<Inputs> = (data) => {
        data["id"] = task.id
        data["priority"] = priority
        data["status_id"] = column!.id
        if (isAuth) {
            if (
                prevTaskObj[task.id].priority !== priority ||
                prevTaskObj[task.id].content !== data.content
            ) {
                dispatch(changeTaskContentQuery(data))
            } else {
                handleClose()
            }
        } else {
            dispatch(changeTaskContent(data))
            handleClose()
        }
    }

    React.useEffect(() => {
        if (successTasksAfterSagaRequest.includes(task.id)) {
            handleClose()
            dispatch(deleteTaskInSuccessArray(task.id))
        }
    }, [successTasksAfterSagaRequest])

    const handleClose = () => {
        setEditMod(false)
        dispatch(deleteTaskInEditArray(task.id)) // удалить таску из (массива с тасками) которые редактируются
    }

    return (
        <form onSubmit={handleSubmit(onSubmit)}>
            <AnimatePresence>
                {editMode ? (
                    <motion.div
                        initial={{ height: 0 }}
                        animate={{ height: "auto" }}
                        exit={{ height: 0 }}
                        style={{ overflow: "hidden" }}
                        // transition={{ duration: 2 }}
                    >
                        <TextareaAutosize
                            className={
                                editMode
                                    ? "block__content_input editMode"
                                    : "block__content_input"
                            }
                            defaultValue={task.content}
                            autoFocus
                            onFocus={moveCaretAtEnd}
                            autoComplete="off"
                            {...register("content", {
                                required: "cannot be empty",
                            })}
                        />
                    </motion.div>
                ) : (
                    <p className="block__content_text">{task.content}</p>
                )}
            </AnimatePresence>

            <div className="error__message">
                {errors?.content && (
                    <p className="error__message_text tasks">
                        {errors?.content?.message}
                    </p>
                )}
            </div>

            <AnimatePresence>
                {editMode && (
                    <motion.div
                        initial={{ height: 0 }}
                        animate={{ height: "auto" }}
                        exit={{ height: 0 }}
                        style={{ overflow: "hidden" }}
                    >
                        <TasksContentEditMode
                            task={task}
                            priorityArray={priorityArray}
                            isValid={isValid}
                            column={column}
                        />
                    </motion.div>
                )}
            </AnimatePresence>
        </form>
    )
}

export default TasksContent
