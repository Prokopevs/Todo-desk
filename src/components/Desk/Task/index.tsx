import React from "react"
import { pen } from "../../../pictures"
import { addTaskInEditArray } from "../../../Store/reducers/editModeSlice"
import { useAppDispatch } from "../../../hooks/redux"
import TasksContent from "./TaskContent"
import { ITasksProps } from "../../../models/ITasksProps"

const Task: React.FC<ITasksProps> = React.memo(
    ({ task, index, priorityArray, column }) => {
        const dispatch = useAppDispatch()

        const color = priorityArray[task.priority - 1].color
        const [editMode, setEditMod] = React.useState(false)

        const onEditClick = () => {
            setEditMod(true)
            dispatch(addTaskInEditArray(task.id))
        }

        return (
            <>
                <button
                    className={
                        editMode ? "block__inner-editPan active" : "block__inner-editPan"
                    }
                    onClick={onEditClick}
                >
                    <img
                        className="block__inner-editPan-img"
                        src={String(pen)}
                        alt=""
                    ></img>
                </button>

                <div className="block__inner_content">
                    <div
                        className={
                            editMode ? "block__content editMode" : "block__content"
                        }
                    >
                        <div className={editMode ? `` : `pretty__line ${color}`}></div>
                        <div className="block__content_text_area">
                            <TasksContent
                                task={task}
                                editMode={editMode}
                                priorityArray={priorityArray}
                                column={column}
                                setEditMod={setEditMod}
                            />
                        </div>
                    </div>
                </div>
            </>
        )
    }
)

export default Task
