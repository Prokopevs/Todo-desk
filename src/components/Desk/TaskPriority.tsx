import React from "react"
import { useAppDispatch } from "../../hooks/redux"
import { ITasksPrioritySelect } from "../../models/ITasksContent"
import { setOpenPriorityСolumn } from "../../Store/reducers/dndSlice"
import Priority from "../Priority"

const TaskPriority: React.FC<ITasksPrioritySelect> = ({ task, editMode, priorityArray }) => {
    const dispatch = useAppDispatch()

    return (
        <>
            {editMode && (
                <div>
                    {task.isOpen ? (
                        <ul>
                            <Priority priorityArray={priorityArray} task={task} />
                        </ul>
                    ) : (
                        <div>
                            <div className="block__line block__line-task"></div>
                            <div className="modalWindow__button task">
                                <button
                                    className={`block__button task ${
                                        priorityArray[task.priority].color
                                    }`}
                                    onClick={() =>
                                        dispatch(setOpenPriorityСolumn(task.id))
                                    }
                                >
                                    <p className="block__button_text">
                                        {priorityArray[task.priority].description}
                                    </p>
                                </button>
                            </div>
                        </div>
                    )}
                </div>
            )}
        </>
    )
}

export default TaskPriority
