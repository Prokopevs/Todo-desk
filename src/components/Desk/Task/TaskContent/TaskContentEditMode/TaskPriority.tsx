import React from "react"
import { useAppDispatch } from "../../../../../hooks/redux"
import { ITasksContentEditMode } from "../../../../../models/Task/ITasksContent"
import { setOpenPriorityСolumn } from "../../../../../Store/reducers/dndSlice"
import Priority from "./Priority"

const TaskPriority: React.FC<ITasksContentEditMode> = ({ task, priorityArray }) => {
    const dispatch = useAppDispatch()

    return (
        <div>
            {task.isOpen ? (
                <ul>
                    <Priority priorityArray={priorityArray} task={task} />
                </ul>
            ) : (
                <div>
                    <div className="block__line block__line-task"></div>
                    <div className="modalWindow__button task">
                        <button
                            className={`block__button task ${
                                priorityArray[task.priority - 1].color
                            }`}
                            onClick={() => dispatch(setOpenPriorityСolumn(task.id))}
                        >
                            <p className="block__button_text">
                                {priorityArray[task.priority - 1].description}
                            </p>
                        </button>
                    </div>
                </div>
            )}
        </div>
    )
}

export default TaskPriority
