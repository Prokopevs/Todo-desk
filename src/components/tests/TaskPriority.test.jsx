import React from "react"
import { render, screen, fireEvent } from "@testing-library/react"
import priorityArray from "../../data/Desk/priorityArray"
import Task from "../Desk/Task"
import { MemoryRouter } from "react-router-dom"
import { Provider } from "react-redux"
import { store } from "../../Store/store"
import "@testing-library/jest-dom"
import * as reactHooks from "../../hooks/redux"
import TaskPriority from "../Desk/Task/TaskContent/TaskContentEditMode/TaskPriority"

const TaskObj = {
    id: "1",
    content: "Watch my favorite",
    priority: 1,
    isOpen: false,
}

jest.mock("../../hooks/redux")

const mockedDispatch = jest.spyOn(reactHooks, "useAppDispatch")

beforeEach(() => {
    mockedDispatch.mockClear()
})

describe("TaskPriority", () => {
    it("should create TaskSelect component", () => {
        mockedDispatch.mockReturnValue(jest.fn())

        const component = render(
            <Provider store={store}>
                <MemoryRouter>
                    <TaskPriority task={TaskObj} priorityArray={priorityArray} />
                </MemoryRouter>
            </Provider>
        )

        expect(component).toMatchSnapshot()
    });

    it("EditMode click event", () => {
        mockedDispatch.mockReturnValue(jest.fn())

        render(
            <Provider store={store}>
                <MemoryRouter>
                    <TaskPriority task={TaskObj} priorityArray={priorityArray} />
                </MemoryRouter>
            </Provider>
        )

        const btn = screen.getByTestId("toggle-btn")
        fireEvent.click(btn)
        expect(mockedDispatch).toHaveBeenCalledTimes(1)
    })
})
