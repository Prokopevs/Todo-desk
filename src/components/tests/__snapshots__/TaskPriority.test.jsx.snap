// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskPriority should create TaskSelect component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div>
          <div
            class="block__line block__line-task"
          />
          <div
            class="modalWindow__button task"
          >
            <button
              class="block__button task purple"
              data-testid="toggle-btn"
            >
              <p
                class="block__button_text"
              >
                Very High
              </p>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        <div
          class="block__line block__line-task"
        />
        <div
          class="modalWindow__button task"
        >
          <button
            class="block__button task purple"
            data-testid="toggle-btn"
          >
            <p
              class="block__button_text"
            >
              Very High
            </p>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
